x/shield/keeper/withdraw.go

```diff
@@ -105,3 +110,225 @@ func (k Keeper) DequeueCompletedWithdrawQueue(ctx sdk.Context) {
 	k.SetTotalCollateral(ctx, totalCollateral)
 	k.SetTotalWithdrawing(ctx, totalWithdrawing)
 }
+
+// ComputeWithdrawAmountByTime computes the amount of collaterals
+// from a given provider that will be dequeued from the withdraw
+// queue by a given time.
+func (k Keeper) ComputeWithdrawAmountByTime(ctx sdk.Context, provider sdk.AccAddress, time time.Time) sdk.Int {
+	withdrawTimesliceIterator := k.WithdrawQueueIterator(ctx, time)
+	defer withdrawTimesliceIterator.Close()
+
+	amount := sdk.ZeroInt()
+	for ; withdrawTimesliceIterator.Valid(); withdrawTimesliceIterator.Next() {
+		timeslice := []types.Withdraw{}
+		value := withdrawTimesliceIterator.Value()
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(value, &timeslice)
+
+		for _, withdraw := range timeslice {
+			if withdraw.Address.Equals(provider) {
+				amount = amount.Add(withdraw.Amount)
+			}
+		}
+	}
+	return amount
+}
+
+func (k Keeper) ComputeTotalUnbondingAmount(ctx sdk.Context, provider sdk.AccAddress) sdk.Int {
+	unbondings := k.sk.GetAllUnbondingDelegations(ctx, provider)
+
+	sum := sdk.ZeroInt()
+	for _, ubd := range unbondings {
+		for i := 0; i < len(ubd.Entries); i++ {
+			sum = sum.Add(ubd.Entries[i].Balance)
+		}
+	}
+	return sum
+}
+
+func (k Keeper) ComputeUnbondingAmountByTime(ctx sdk.Context, provider sdk.AccAddress, time time.Time) sdk.Int {
+	dvPairs := k.GetUnbondingsByProviderMaturingByTime(ctx, provider, time)
+
+	sum := sdk.ZeroInt()
+	var seen []sdk.ValAddress
+	for _, dvPair := range dvPairs {
+		valAddr := dvPair.Validator
+		if find(seen, valAddr) {
+			continue
+		}
+		seen = append(seen, valAddr)
+
+		// obtain unbonding entries and iterate through them
+		ubd, found := k.sk.GetUnbondingDelegation(ctx, provider, valAddr)
+		if !found {
+			continue //TODO
+		}
+		for i := 0; i < len(ubd.Entries); i++ {
+			entry := ubd.Entries[i]
+			if entry.IsMature(time) {
+				sum = sum.Add(entry.Balance)
+			} else {
+				break
+			}
+		}
+	}
+	return sum
+}
+
+func find(list []sdk.ValAddress, item sdk.ValAddress) bool {
+	for _, val := range list {
+		if val.Equals(item) {
+			return true
+		}
+	}
+	return false
+}
+
+type UnbondingInfo struct {
+	Delegator sdk.AccAddress 
+	Validator sdk.ValAddress
+	CompletionTime time.Time
+}
+
+func (k Keeper) GetUnbondingsByProviderMaturingByTime(ctx sdk.Context, provider sdk.AccAddress, time time.Time) (results []UnbondingInfo) {
+	unbondingTimesliceIterator := k.sk.UBDQueueIterator(ctx, ctx.BlockHeader().Time)
```

Comment:
Should `ctx.BlockHeader().Time` be `time`?

Reply:
yes thanks :)